Recent failures in ExternalDriverAPIWeakCplDAProto build on Darwin machines in debug mode

Recent Darwin testing in debug mode (g) has been experiencing failures in the ExternalDriverAPIWeakCplDAProto build (on green, grits and hearhear), with a build error like this:

```
mpicxx -pthread -L/Users/sacks/scratch/esmf-testing/gfortranclang_14.1.0_16.0.0_openmpi_g_develop/esmf/DEFAULTINSTALLDIR/lib/libg/Darwin.gfortranclang.64.openmpi.default -L/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0-gfortran14.1.0/netcdf-c-4.9.2-qgk7t4q3nzy2b6jminptofqpvku7uefb/lib -L/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0-gfortran14.1.0/netcdf-fortran-4.6.1-4372uma2tlhbvvffzy2lnurayzjyx6fk/lib -L/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0/gcc-14.1.0-yxln5vjswuutfxtzpq4tf5yk7iuhin4a/lib/gcc/aarch64-apple-darwin23.6.0/14.1.0/../../../ -L./ -Wl,-rpath,/Users/sacks/scratch/esmf-testing/gfortranclang_14.1.0_16.0.0_openmpi_g_develop/esmf/DEFAULTINSTALLDIR/lib/libg/Darwin.gfortranclang.64.openmpi.default -Wl,-rpath,/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0-gfortran14.1.0/netcdf-c-4.9.2-qgk7t4q3nzy2b6jminptofqpvku7uefb/lib -Wl,-rpath,/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0-gfortran14.1.0/netcdf-fortran-4.6.1-4372uma2tlhbvvffzy2lnurayzjyx6fk/lib  -Wl,-rpath,/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0/gcc-14.1.0-yxln5vjswuutfxtzpq4tf5yk7iuhin4a/lib/gcc/aarch64-apple-darwin23.6.0/14.1.0/../../../ -o externalApp externalApp.o atmDA.o ocnDA.o nuopc_da.o esm.o atm.o ocn.o -lesmf -lnetcdf -lnetcdff -lnetcdf -lnetcdf -lm -lpioc -L/Users/sacks/spack/opt/spack/darwin-sonoma-m2/apple-clang-16.0.0-gfortran14.1.0/openmpi-5.0.5-en3fy3sfg6du2nqnfhtapxsrzmmiakzr/lib -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi -lm -lc++ -lgfortran
ld: warning: ignoring duplicate libraries: '-lc++', '-lm', '-lmpi', '-lnetcdf'
Undefined symbols for architecture arm64:
  "___gcc_nested_func_ptr_created", referenced from:
      ___nuopc_da_MOD_commtovm in nuopc_da.o
  "___gcc_nested_func_ptr_deleted", referenced from:
      ___nuopc_da_MOD_commtovm in nuopc_da.o
ld: symbol(s) not found for architecture arm64
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [externalApp] Error 1
```

(That is on hearhear. On green, the error is similar, but with `builtin_nested_...` instead of `gcc_nested_...`. I haven't checked grits.)

The issue is the nested recursive subroutine in nuopc_da.F90. My best guess is that the recent OS update on Macs has caused this issue due to the `nested_func_ptr` routines no longer being present in the Apple clang standard library. (I saw a description somewhere that this may only be relevant on older architectures.) But it's odd that this works in O builds.

In any case, there is a simple workaround that I will submit shortly.

FYI @theurich @oehmke .

--- a/ExternalDriverAPIWeakCplDAProto/nuopc_da.F90
+++ b/ExternalDriverAPIWeakCplDAProto/nuopc_da.F90
@@ -116,6 +116,15 @@ module nuopc_da
 
   !-----------------------------------------------------------------------------
 
+  ! The following is used in commToVM. Some Mac configurations have issues with this being
+  ! an internal subroutine to commToVM, so we have pulled it out to a module-level
+  ! subroutine. (See https://github.com/esmf-org/nuopc-app-prototypes/issues/8.)
+  recursive subroutine dummySS(gridcomp, rc)
+    type(ESMF_GridComp)        :: gridcomp ! must not be optional
+    integer, intent(out)       :: rc       ! must not be optional
+    rc = ESMF_SUCCESS
+  end subroutine
+
   function commToVM(comm, rc)
     type(ESMF_VM) :: commToVM
     integer,       intent(in)   :: comm ! MPI communicator
@@ -187,14 +196,6 @@ module nuopc_da
     if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
       line=__LINE__, &
       file=__FILE__)) return
-
-  contains
-
-    recursive subroutine dummySS(gridcomp, rc)
-      type(ESMF_GridComp)        :: gridcomp ! must not be optional
-      integer, intent(out)       :: rc       ! must not be optional
-      rc = ESMF_SUCCESS
-    end subroutine
 
   end function commToVM
 

